File: test/python/test_rticonnextdds_output.py
import pytest,sys,os
sys.path.append(os.path.dirname(os.path.realpath(__file__))+ "/../../")
import rticonnextdds_connector as rti

class TestOutput:

  def test_invalid_DW(self,rtiConnectorFixture):
    invalid_DW = "InvalidDW"
    with pytest.raises(ValueError):
      op= rtiConnectorFixture.getOutput(invalid_DW)
 
  def test_creation_DW(self,rtiOutputFixture):
    assert isinstance(rtiOutputFixture,rti.Output) \
      and rtiOutputFixture.name == "MyPublisher::MySquareWriter" \
      and isinstance(rtiOutputFixture.connector,rti.Connector) \
      and isinstance(rtiOutputFixture.instance, rti.Instance)
    
class TestInstance:

  def test_instance_creation(self,rtiOutputFixture):
    assert rtiOutputFixture.instance!=None and \
      isinstance(rtiOutputFixture.instance, rti.Instance)

  # TODO: No exception is thrown when a non-existent field is accessed
  """
  An AttributeError must be propagated on setNumber,setString & setBoolean
  KeyError must be propagated on setDictionary with invalid fields 
  """
  @pytest.mark.xfail
  def test_setNumber_on_nonexistent_field(self,rtiOutputFixture,capfd):
    non_existent_field="invalid_field"
    rtiOutputFixture.instance.setNumber(non_existent_field,1)
    out,err = capfd.readouterr()
    assert "RTILuaDynamicData_set:!get kind failed" in out

  @pytest.mark.xfail
  def test_setString_on_nonexistent_field(self,rtiOutputFixture,capfd):
    non_existent_field="invalid_field"
    rtiOutputFixture.instance.setString(non_existent_field,"1")
    out,err = capfd.readouterr()
    assert "RTILuaDynamicData_set:!get kind failed" in out

  @pytest.mark.xfail
  def test_setBoolean_on_nonexistent_field(self,rtiOutputFixture,capfd):
    non_existent_field="invalid_field"
    rtiOutputFixture.instance.setBoolean(non_existent_field,True)
    out,err = capfd.readouterr()
    assert "RTILuaDynamicData_set:!get kind failed" in out

  @pytest.mark.xfail
  def test_setDictionary_with_nonexistent_fields(self,rtiOutputFixture,capfd):
    invalid_dictionary= {"non_existent_field":"value"}
    rtiOutputFixture.instance.setDictionary(invalid_dictionary)
    out,err = capfd.readouterr()
    assert "RTILuaJsonHelper_parse_json_node:!get kind failed" in out

  @pytest.mark.xfail
  # Implicit type conversion from Boolean to number 
  def test_setNumber_with_Boolean(self,rtiOutputFixture):
    number_field="x"
    with pytest.raises(TypeError) as execinfo:
      rtiOutputFixture.instance.setNumber(number_field,True)
    print("\nException of type:"+str(execinfo.type)+ \
      "\nvalue:"+str(execinfo.value))

  def test_setNumber_with_String(self,rtiOutputFixture):
    number_field="x"
    with pytest.raises(TypeError) as execinfo:
      rtiOutputFixture.instance.setNumber(number_field,"str")
    print("\nException of type:"+str(execinfo.type)+ \
      "\nvalue:"+str(execinfo.value))
 
  def test_setNumber_with_Dictionary(self,rtiOutputFixture):
    number_field="x"
    with pytest.raises(TypeError) as execinfo:
      rtiOutputFixture.instance.setNumber(number_field,{"x":1})
    print("\nException of type:"+str(execinfo.type)+ \
      "\nvalue:"+str(execinfo.value))

  def test_setString_with_Boolean(self,rtiOutputFixture):
    string_field="color"
    with pytest.raises(AttributeError) as execinfo:
      rtiOutputFixture.instance.setString(string_field,True)
    print("\nException of type:"+str(execinfo.type)+ \
      "\nvalue:"+str(execinfo.value))

  def test_setString_with_Number(self,rtiOutputFixture):
    string_field="color"
    with pytest.raises(AttributeError) as execinfo:
      rtiOutputFixture.instance.setString(string_field,55.55)
    print("\nException of type:"+str(execinfo.type)+ \
      "\nvalue:"+str(execinfo.value))
  
  def test_setString_with_Dictionary(self,rtiOutputFixture):
    string_field="color"
    with pytest.raises(AttributeError) as execinfo:
      rtiOutputFixture.instance.setString(string_field,{"color":1})
    print("\nException of type:"+str(execinfo.type)+ \
      "\nvalue:"+str(execinfo.value))


  def test_setBoolean_with_String(self,rtiOutputFixture):
    boolean_field="z"
    with pytest.raises(TypeError) as execinfo:
      rtiOutputFixture.instance.setBoolean(boolean_field,"str")
    print("\nException of type:"+str(execinfo.type)+ \
      "\nvalue:"+str(execinfo.value))

  @pytest.mark.xfail
  # Implicit type conversion from number to Boolean 
  def test_setBoolean_with_Number(self,rtiOutputFixture):
    boolean_field="z"
    with pytest.raises(TypeError) as execinfo:
      rtiOutputFixture.instance.setBoolean(boolean_field,55.55)
    print("\nException of type:"+str(execinfo.type)+ \
      "\nvalue:"+str(execinfo.value))
  
  def test_setBoolean_with_Dictionary(self,rtiOutputFixture):
    boolean_field="z"
    with pytest.raises(TypeError) as execinfo:
      rtiOutputFixture.instance.setBoolean(boolean_field,{"color":1})
    print("\nException of type:"+str(execinfo.type)+ \
      "\nvalue:"+str(execinfo.value))

  # TODO: A dictionary with incompatible types can be set!!!
  @pytest.mark.xfail
  def test_setDictionary_with_incompatible_types(self,rtiOutputFixture,capfd):
    dict_with_incompatible_types={"color":1,"x":"str"}
    rtiOutputFixture.instance.setDictionary(dict_with_incompatible_types)
    out,err = capfd.readouterr()
    assert 0

File: test/nodejs/test_rticonnextdds_output.js
var expect=require('chai').expect
var rti= require(__dirname + '/../../rticonnextdds-connector')

describe('Output Tests',function() {
  var connector=null 
  // Initialization before all tests execute
  before(function(){
    var participant_profile = "MyParticipantLibrary::Zero"
    var xml_profile = __dirname + "/../xml/ShapeExample.xml"
    connector = new rti.Connector(participant_profile,xml_profile)
  })

  // Cleanup after all tests have executed
  after(function() {
    this.timeout(0)
    connector.delete()
    
  })

  it('Output object should not get instantiated for invalid DataWriter',function(){
    var invalid_DW = "invalidDW"
    expect(function(){
      connector.getWriter(invalid_DW)
    }).to.throw(Error)
  }) 

  it('Output object should get instantiated for valid ' +
      'Publication::DataWriter name',function (){
    var valid_DW = "MyPublisher::MySquareWriter" 
    var output= connector.getOutput(valid_DW)
    expect(output).to.exist 
    expect(output.name).to.equal(valid_DW)
    expect(output.connector).to.equal(connector)
  })

  describe('Tests on Output\'s Instance',function() {
    var output = null
    //Initialization before all tests execute in this describe block
    before(function(){
      output= connector.getOutput("MyPublisher::MySquareWriter")  
    })

    it('output\'s instance should exist',function(){
      expect(output.instance).to.exist
    })

    //skipped test: Condition being tested has not been accounted for yet
    it.skip('setNumber on non-existent field should throw error and '+
      'subscriber should not get a message with default values',function(){
      expect(function (){
        output.instance.setNumber("invalid_field",1)
      }).to.throw(Error)
    })

    //skipped test: Condition being tested has not been accounted for yet
    it.skip('setString on non-existent field should throw error and '+
      'subscriber should not get a message with default values',function(){
      expect(function (){
        output.instance.setString("invalid_field","value")
      }).to.throw(Error)
    })

    //skipped test: Condition being tested has not been accounted for yet
    it.skip('setBoolean on non-existent field should throw error and '+
      'subscriber should not get a message with default values',function(){
      expect(function (){
        output.instance.setBoolean("invalid_field",true)
      }).to.throw(Error)
    })

    //skipped test: Condition being tested has not been accounted for yet
    it.skip('setFromJSON should throw error for a JSON object ' +
      'with non-existent fields and subscriber should not get ' +
      'a message with default values',function(){
       expect(function (){
          var invalid_data = '{"invalid_field":1}'
          output.instance.setFromJSON(JSON.parse(
            '{"invalid_field":1}'))
      }).to.throw(Error)
    })

    it('setString with boolean value should throw Error', function(){
      expect(function(){
        var string_field="color"
        output.instance.setString(string_field,true)
      }).to.throw(Error)
    })

    it('setString with number value should throw Error', function(){
      expect(function(){
        var string_field="color"
        output.instance.setString(string_field,11)
      }).to.throw(Error)
    })

    it('setString with dictionary value should throw Error', function(){
      expect(function(){
        var string_field="color"
        output.instance.setString(string_field,{"key":"value"})
      }).to.throw(Error)
    })

    //skipped test: Condition being tested has not been accounted for yet
    it.skip('setNumber with string value should throw Error and'+
      'subscriber should not get a message with erroneous field data', function(){
      expect(function(){
        var number_field="x"
        output.instance.setNumber(number_field,"value")
      }).to.throw(Error)
    })

    // un-implemented test
    it('Note: implicit type-conversion for setNumber with boolean value')

    //skipped test: Condition being tested has not been accounted for yet
    it.skip('setNumber with dictionary value should throw Error and '+
      'subscriber should not get a message with erroneous field data', function(){
      expect(function(){
        var number_field="x"
        output.instance.setNumber(number_field,{"key":"value"})
      }).to.throw(Error)
    })
 
    //skipped test: Condition being tested has not been accounted for yet
    it.skip('setBoolean with string value should throw Error and '+
      'subscriber should not get a  message with erroneous field data', function(){
      expect(function(){
        var boolean_field="z"
        output.instance.setBoolean(boolean_field,"value")
      }).to.throw(Error)
    })

    // unimplemented test
    it('Note: implicit type-conversion for setBoolean with number value')

    //skipped test: Condition being tested has not been accounted for yet
    it.skip('setBoolean with dictionary value should throw Error and '+
      'subscriber should not get a  message with erroneous field data', function(){
      expect(function(){
        var boolean_field="z"
        output.instance.setBoolean(boolean_field,{"key":"value"})
      }).to.throw(Error)
    })

    //skipped test: Condition being tested has not been accounted for yet
    it.skip('setFromJSON for JSON object with incompatible value types ' + 
      'should throw Error and subscriber should not get a message with ' + 
      'erroneous field data', function(){
        expect(function(){
          var str= '{"x":"5","y":true,"color":true,"shapesize":"5","z":"value"}'
          output.instance.setFromJSON(JSON.parse(str))
        }).to.throw(Error)
    })
  })

})
